PurchasesCoreSwift/SubscriberAttributes/SubscriberAttributesManager.swift

```diff
@@ -0,0 +1,261 @@
+        setAttribute(key: SpecialSubscriberAttributes.ad, value: maybeAd, appUserID: appUserID)
+    }
+
+    @objc public func setKeyword(_ maybeKeyword: String?, appUserID: String) {
+        Logger.debug(String(format: Strings.attribution.method_called, "setKeyword"))
+        setAttribute(key: SpecialSubscriberAttributes.keyword, value: maybeKeyword, appUserID: appUserID)
+    }
+
+    @objc public func setCreative(_ maybeCreative: String?, appUserID: String) {
+        Logger.debug(String(format: Strings.attribution.method_called, "setCreative"))
+        setAttribute(key: SpecialSubscriberAttributes.creative, value: maybeCreative, appUserID: appUserID)
+    }
+
+    @objc public func collectDeviceIdentifiers(forAppUserID appUserID: String) {
+        Logger.debug("collectDeviceIdentifiers called")
+        // TODO safe to remove log below?
+        Logger.debug(String(format: Strings.attribution.method_called, "setAttributes"))
+        let identifierForAdvertisers = attributionFetcher.identifierForAdvertisers
+        let identifierForVendor = attributionFetcher.identifierForVendor
+
+        setAttribute(key: SpecialSubscriberAttributes.idfa, value: identifierForAdvertisers, appUserID: appUserID)
+        setAttribute(key: SpecialSubscriberAttributes.idfv, value: identifierForVendor, appUserID: appUserID)
+        setAttribute(key: SpecialSubscriberAttributes.ip, value: "true", appUserID: appUserID)
+    }
+
+    @objc public func syncAttributesForAllUsers(currentAppUserID: String) {
+        let unsyncedAttributesForAllUsers = unsyncedAttributesByKeyForAllUsers()
+
+        for (syncingAppUserId, attributes) in unsyncedAttributesForAllUsers {
+            syncAttributes(attributes: attributes, appUserID: syncingAppUserId) { error  in
+                self.handleAttributesSynced(syncingAppUserId: syncingAppUserId,
+                                            currentAppUserId: currentAppUserID,
+                                            error: error)
+            }
+        }
+    }
+
+    @objc public func handleAttributesSynced(syncingAppUserId: String, currentAppUserId: String, error: Error?) {
+        if error == nil {
+            Logger.rcSuccess(String(format: Strings.attribution.attributes_sync_success, syncingAppUserId))
+            if syncingAppUserId != currentAppUserId {
+                deviceCache.deleteAttributesIfSynced(appUserID: syncingAppUserId)
+            }
+        } else {
+            let receivedNSError = error as NSError?
+            Logger.error(String(format: Strings.attribution.attributes_sync_error,
+                                receivedNSError?.localizedDescription ?? "",
+                                receivedNSError?.userInfo ?? ""))
+        }
+    }
+
+    @objc public func unsyncedAttributesByKey(appUserID: String) -> SubscriberAttributeDict {
+        // TODO  these logs were in RCPurchases+subscriberattributes, are we ok with logging
```

Comment:
another question here in the TODO...

Reply:
I'm not familiar enough to know for sure, but if it contains any sensitive information at all, I'd say no, we shouldn't log it. 
Also, we shouldn't it's not useful for customer debugging or support usage.

Reply:
@aboedo mind weighing in on this one?

Reply:
They're not supposed to be used for sensitive information, [as stated in the docs](https://docs.revenuecat.com/v3.1/docs/subscriber-attributes#setting-attributes). 

> Since subscriber attributes are writable using a public key they should not be used for managing secure or sensitive information such as subscription status, coins, etc.

That said, I think the more valuable part of the logs is just logging aggregated counts, or at most just the keys, the values aren't useful at all. 

As for usefulness for support, we have had quite a bit of customer confusion because devs set something and it doesn't show up right away in the dashboard, so logs do come into play for that kind of thing. Or when overwriting values. 

That said, we might still be able to get value out of the logs without logging out the value of the attributes. 
Come to think about it, we might wanna add a log saying something to the effect of: "attributes set, they'll be synced to the backend when the app goes into background / foreground or when a purchase is made"

Reply:
@aboedo is the appUserId not considered sensitive?

Reply:
Without knowing of any previous discussions around PII (personally identifiable information if you're not familiar with that acronym), my take which I adopted from my team at Google is- Basically anything that could remotely be tracked back to a human is PII and therefore, should be treated as sensitive. For my previous projects we used hashes for things if we needed stats or entire session dumps. Same things goes for API endpoints, we always used specially generated "external ids" that were safer than userIds.

Reply:
we encourage devs to use non-guessable, non-sensitive (ideally UUID) appUserIDs. 

We can't guarantee that they won't, though, so I'm not opposed to avoiding it in logs. Perhaps we could tie it to the log level (i.e.: only print it in debug logs, but not info or above). 

Note that we can actually ensure that debug logs aren't used in production by doing something like 

`Logger.swift`
```swift
    @objc public static func log(level: LogLevel, message: String) {
        guard self.logLevel.rawValue <= level.rawValue else { return }
#if !DEBUG
        guard logLevel > .debug else { return }
#endif
        logHandler(level, message)
    }
```

Reply:
+1 on that idea ^^ We do seem to pass userID to a lot of our logs, though...perhaps I can create a follow-up ticket?

Reply:
sounds good. note that we were doing this _before_ the migration, so it's not strictly a migration task